File: offer/simple/PHP/mergeTwoLists.php
<?php

/**
 * 剑指 Offer 25. 合并两个排序的链表
 * @Author: pengwei
 * @Date:   2021-05-22 19:07:33
 * @Last Modified by:   pengwei
 */
function mergeTwoLists($l1, $l2) {
    $l3 = new ListNode(null);
    $head = $l3;

    while ($l1 && $l2) {
        if ($l1->val <= $l2->val) {
            $nodeVal = $l1->val;
            $l1 = $l1->next;
        } elseif ($l1->val > $l2->val) {
            $nodeVal = $l2->val;
            $l2 = $l2->next;
        }
        
       $l3->next = new ListNode($nodeVal);
       $l3 = $l3->next;
    }
    
    while ($l1) {
        $nodeVal = $l1->val;
        $l1 = $l1->next;
        $l3->next = new ListNode($nodeVal);
        $l3 = $l3->next;
    }
    
    while ($l2) {
        $nodeVal = $l2->val;
        $l2 = $l2->next;
        $l3->next = new ListNode($nodeVal);
        $l3 = $l3->next;
    }
    return $head->next;
}

File: offer/simple/Golang/mergeTwoLists.go
// 剑指 Offer 25. 合并两个排序的链表
// 输入两个递增排序的链表，合并这两个链表并使新链表中的节点仍然是递增排序的。

package main

//type ListNode struct {
//	val int
//	next *ListNode
//}

func mergeTwoLists(l1 *listNode, l2 *listNode) *listNode {
	node := &listNode{}
	curr := node

	for l1 != nil && l2 != nil {
		if l1.val <= l2.val {
			curr.next = l1
			l1 = l1.next
		} else {
			curr.next = l2
			l2 = l2.next
		}
		curr = curr.next
	}

	// 两链表是否存在未比较完的元素
	if l1 != nil {
		curr.next = l1
	}

	if l2 != nil {
		curr.next = l2
	}

	return node.next
}